cmd/kubeadm/app/util/pkiutil/pki_helpers.go

```diff
@@ -256,6 +292,21 @@ func TryLoadCertFromDisk(pkiPath, name string) (*x509.Certificate, error) {
 	return cert, nil
 }
 
+// TryLoadCertChainFromDisk tries to load the cert chain from the disk
+func TryLoadCertChainFromDisk(pkiPath, name string) (*x509.Certificate, []*x509.Certificate, error) {
+	certificatePath := pathForCert(pkiPath, name)
+
+	certs, err := certutil.CertsFromFile(certificatePath)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "couldn't load the certificate file %s", certificatePath)
+	}
+
+	cert := certs[0]
+	intermediates := certs[1:]
```

Comment:
AFAIK, orders in chains are not guaranteed, thus this feels applicable to your use case only.

Reply:
if a chain contains no intermediates, this line will panic.

Reply:
> AFAIK, orders in chains are not guaranteed, thus this feels applicable to your use case only.

I can't find anything explicit in [RFC 5280](https://tools.ietf.org/html/rfc5280) or [RFC 7468](https://tools.ietf.org/html/rfc7468) to say that the leaf certificate must be first. However, I can see that the `crypto/tls` library makes this assumption in its [certificate verification logic](https://github.com/golang/go/blob/663cd862edf8dfa9c78d9df0f372c9bea03139e4/src/crypto/tls/handshake_client.go#L851) as well an explicit comment elsewhere that the [leaf certificate should be first](https://github.com/golang/go/blob/663cd862edf8dfa9c78d9df0f372c9bea03139e4/src/crypto/tls/common.go#L1323).

Since I can't find an explicit statement in the standards I'm going to have to fall back on the argument that if it's good enough for Golang's standard TLS implementation then it's good enough for the certificate verification helpers in kubeadm. The Kubernetes components will be using the TLS implementation anyway, so it makes sense that the kubeadm phases which perform certificate verification behave the same way.

Reply:
> if a chain contains no intermediates, this line will panic.

It shouldn't panic, if there's only a single certificate then `intermediates` will be an empty slice. Here's a quick example:

```go
package main

import "fmt"

func main() {
    x := []int{1}
    fmt.Printf("Head: %d\n", x[0])
    fmt.Printf("Tail: %v\n", x[1:])
}
```

This produces:

```
Head: 1
Tail: []
```

Reply:
[RFC5286 7.4.2](https://tools.ietf.org/html/rfc5246) describes the chain order:

> The sender's certificate MUST come first in the list.  Each following certificate MUST directly certify the one preceding it.  Because certificate validation requires that root keys be distributed independently, the self-signed certificate that specifies the root certificate authority MAY be omitted from the chain, under the assumption that the remote end must already possess it in order to validate it in any case.

Reply:
> It shouldn't panic,

my mistake.

> RFC5286 7.4.2 describes the chain order:

thanks for the info!